### ISSUE TITLE
About 'Clone' Function

### ISSUE THREAD

**Issue by whzyf951620** [learnables/learn2learn#169]
For example, in the examples/vision/maml_miniimagenet.py, learner = maml.clone() is normal copy or deepcopy?
In training process, the gradients are computed by  evaluation_error.backward().
Do they be saved in the parameters of maml.model or learner which is the copy of maml.model?
Then, in the validation process, another 'learner' is yielded by  maml.clone() and it is adapted quickly in validation process.
Will it change the parameters or their gradients of maml.model?
Finally, opt.step() is used to update the parameters of maml.model, which gradients are used?

I am looking forward to your return, 
Thank you very much!

**Comment by whzyf951620** [comment]
The training of maml has not update the parameters of maml until the opt.step()?
The process of fast adaptation is only used to compute the appropriate gradients of the final update?

**Comment by seba-1511** [comment]
Hello @whzyf951620 ,

`maml.clone()` internally calls `clone_module()` which you can think of as a deepcopy that retains the computational graph dependencies between the original module and the clone. So when we adapt the clone, the original module remains untouched. But, when call `loss.backward()` on a loss computed with the clone, you're computing the gradients w.r.t. the parameters of the original module. (i.e. the backpropagation "flows through" the adaptation and clone steps.)

For more information, feel free to refer to the documentation of `clone_module`: http://learn2learn.net/docs/learn2learn/#clone_module

I hope this helps.

**Comment by whzyf951620** [comment]
When I want to get your performance on Mini-ImageNet published on 'http://learn2learn.net/examples.vision/' by your examples, I found that your meta lr and fast lr are very different from the lr values in MAML papers. Also I cannot use your 
examlpe codes 'maml_omniglot.py' and 'maml_miniimagenet' to get your test acc.

Can you guide me? 
Thank you very much and I am looking forward to your return.

**Comment by seba-1511** [comment]
I think the meta and fast learning rates are different mostly because some implementations average the loss over episodes (e.g. ours) and some don't (e.g. C. Finn's).

As for your questions on the MAML examples, feel free to join our slack and someone might help you there: [slack.learn2learn.net](http://slack.learn2learn.net)

**Comment by seba-1511** [comment]
Inactive - closing.
